source/crc.c

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include "checksums.h"
+
+#include "aws/checksums/crc.h"
+#include "aws/common/byte_buf.h"
+
+PyObject *aws_py_checksums_crc32(PyObject *self, PyObject *args) {
+    (void)self;
+
+    Py_buffer input;
+    PyObject *py_previousCrc32;
+
+    if (!PyArg_ParseTuple(args, "s*O", &input, &py_previousCrc32)) {
+        return NULL;
+    }
+
+    /* Note: PyArg_ParseTuple() doesn't do overflow checking on unsigned values
+     * so use PyLong_AsUnsignedLong() to get the value of the previousCrc arg */
+    uint32_t previousCrc32 = PyLong_AsUnsignedLong(py_previousCrc32);
+
+    if (previousCrc32 == (uint32_t)-1 && PyErr_Occurred()) {
+        return NULL;
+    }
+
+    if (!PyBuffer_IsContiguous(&input, 'C')) {
+        if (input.obj) {
+            PyBuffer_Release(&input);
+        }
+        PyErr_SetString(PyExc_ValueError, "input must be contiguous buffer");
+        return NULL;
+    }
+
+    uint32_t signed_val = previousCrc32;
+
+    /* Releasing the GIL for very small buffers is inefficient
+       and may lower performance */
+    if (input.len > 1024 * 5) {
+        unsigned char *buf = input.buf;
+        Py_ssize_t len = input.len;
+
+        /* clang-format off */
+        Py_BEGIN_ALLOW_THREADS
+            /* Avoid truncation of length for very large buffers. crc32() takes
+               length as an unsigned int, which may be narrower than Py_ssize_t. */
+            while ((size_t)len > UINT_MAX) {
+                signed_val = aws_checksums_crc32(buf, UINT_MAX, signed_val);
+                buf += (size_t)UINT_MAX;
+                len -= (size_t)UINT_MAX;
```

Comment:
aws_checksums_crc32() takes length as an `int`, not a `uint`
~~length as an unsigned int,~~ -> length as a signed int
~~UINT_MAX~~ -> INT_MAX

add a test that hits this branch. It's not correct at the moment, and definitely worth testing given the complexity

I'm actually pretty spooked why the compiler isn't complaining about this

Reply:
for the test, it might very well fail on a computer with insufficient memory, so the code would be something like:
```
# stress the internal logic that handles buffers larger than C's INT_MAX
try:
    INT_MAX = 2**31 - 1
    giant_buffer = bytes(INT_MAX + 5)
except MemoryError:
    raise unittest.SkipTest('Machine can't allocate giant buffer for giant buffer test')

val = crc32(giant_buffer)
```